File: core/src/entities/recipe.ts
/**
 * Recipe-domain entities. Port of `domain/entities/recipe.py`.
 *
 * Pure in-memory shapes the business logic operates on — no ORM, no I/O. Fields
 * that carry defaults in the Python dataclasses are optional here; the storage
 * layer (a later step) is responsible for supplying them on read.
 *
 * Identity is a UUID string (`Id`) — see ../identity.ts. `usdaFdcId` is the one
 * exception: it's an external USDA FoodData Central integer, not our identity.
 */
import type { Decimal } from "decimal.js";

import type { Id } from "../identity";

export enum SourceType {
  WEB = "web",
  VIDEO = "video",
  MANUAL = "manual",
}

export enum RecipeStatus {
  DRAFT = "draft",
  FINALIZED = "finalized",
}

export interface Tag {
  name: string;
  id?: Id | null;
}

/** A canonical, deduplicated ingredient in the user's catalog. */
export interface Ingredient {
  name: string;
  normalizedName: string;
  id?: Id | null;
  usdaFdcId?: number | null;
  defaultUnit?: string | null;
}

/**
 * A single ingredient line on a recipe — the ingredient-level macro link.
 * `usdaFdcId` and `gramWeight` are populated by nutrition resolution and drive
 * per-ingredient macro math.
 */
export interface RecipeIngredient {
  ingredient: Ingredient;
  quantity?: Decimal | null;
  unit?: string | null;
  rawText?: string | null;
  position?: number;
  usdaFdcId?: number | null;
  gramWeight?: Decimal | null;
  notes?: string | null;
  id?: Id | null;
}

export interface Recipe {
  title: string;
  id?: Id | null;
  description?: string | null;
  sourceUrl?: string | null;
  sourceType?: SourceType;
  servings?: number | null;
  rating?: number | null;
  status?: RecipeStatus;
  instructionsMd?: string | null;
  /**
   * Recipe-level macros, entered by hand per serving. These are the primary
   * macro source: when any is set, the MacroCalculator uses them directly
   * instead of summing ingredient-level USDA links.
   */
  caloriesPerServing?: Decimal | null;
  proteinPerServing?: Decimal | null;
  fatPerServing?: Decimal | null;
  carbsPerServing?: Decimal | null;
  fiberPerServing?: Decimal | null;
  sodiumPerServing?: Decimal | null;
  ingredients?: RecipeIngredient[];
  tags?: Tag[];
  createdAt?: string | null;
  updatedAt?: string | null;
}

File: backend/src/recetario/domain/entities/recipe.py
"""Pure domain entities — no framework, no ORM, no I/O.

These are the canonical in-memory representations the business logic operates on.
Repositories map persistence rows to/from these objects.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from enum import Enum


class SourceType(str, Enum):
    WEB = "web"
    VIDEO = "video"
    MANUAL = "manual"


class RecipeStatus(str, Enum):
    DRAFT = "draft"
    FINALIZED = "finalized"


@dataclass
class Tag:
    name: str
    id: int | None = None


@dataclass
class Ingredient:
    """A canonical, deduplicated ingredient in the user's catalog."""

    name: str
    normalized_name: str
    id: int | None = None
    usda_fdc_id: int | None = None
    default_unit: str | None = None


@dataclass
class RecipeIngredient:
    """A single ingredient line on a recipe — the ingredient-level macro link.

    `usda_fdc_id` and `gram_weight` are populated by the nutrition-resolution step
    (Phase 2/3) and drive per-ingredient macro math.
    """

    ingredient: Ingredient
    quantity: Decimal | None = None
    unit: str | None = None
    raw_text: str | None = None
    position: int = 0
    usda_fdc_id: int | None = None
    gram_weight: Decimal | None = None
    notes: str | None = None
    id: int | None = None


@dataclass
class Recipe:
    title: str
    id: int | None = None
    description: str | None = None
    source_url: str | None = None
    source_type: SourceType = SourceType.MANUAL
    servings: int | None = None
    rating: int | None = None
    status: RecipeStatus = RecipeStatus.DRAFT
    instructions_md: str | None = None
    # Recipe-level macros, entered by hand per serving. These are the primary
    # macro source: when any is set, the MacroCalculator uses them directly
    # instead of summing ingredient-level USDA links (see macro_calculator).
    calories_per_serving: Decimal | None = None
    protein_per_serving: Decimal | None = None
    fat_per_serving: Decimal | None = None
    carbs_per_serving: Decimal | None = None
    fiber_per_serving: Decimal | None = None
    sodium_per_serving: Decimal | None = None
    ingredients: list[RecipeIngredient] = field(default_factory=list)
    tags: list[Tag] = field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None
